tradeexecutor/ethereum/uniswap_v3_live_pricing.py

```diff
@@ -0,0 +1,251 @@
+
+        return TradePricing(
+            price=price,
+            mid_price=mid_price,
+            lp_fee=lp_fee,
+            pair_fee=fees,
+            side=False,
+        )
+        
+
+    def get_buy_price(self,
+                       ts: datetime.datetime,
+                       pair: TradingPairIdentifier,
+                       reserve: Optional[Decimal],
+                       ) -> USDollarAmount:
+        """Get live price on Uniswap.
+
+        :param reserve:
+            The buy size in quote token e.g. in dollars
+
+        :return: Price for one reserve unit e.g. a dollar
+        """
+
+        if reserve is None:
+            reserve = Decimal(self.very_small_amount)
+        else:
+            assert isinstance(reserve, Decimal), f"Reserve must be decimal, got {reserve.__class__}: {reserve}"
+
+        target_pair, intermediate_pair = self.routing_model.route_pair(self.pair_universe, pair)
+
+        base_addr, quote_addr, intermediate_addr = route_tokens(target_pair, intermediate_pair)
+
+        # In three token trades, be careful to use the correct reserve token
+        if intermediate_pair is not None:
+            reserve_raw = intermediate_pair.quote.convert_to_raw_amount(reserve)
+            self.check_supported_quote_token(intermediate_pair)
+        else:
+            reserve_raw = target_pair.quote.convert_to_raw_amount(reserve)
+            self.check_supported_quote_token(pair)
+
+        # See eth_defi.uniswap_v2.fees.estimate_buy_received_amount_raw
+        path = (
+            [quote_addr, intermediate_addr, base_addr]
+            if intermediate_addr
+            else [quote_addr, base_addr] 
+        )
+        fees = (
+            [intermediate_pair.fee, target_pair.fee]
+            if intermediate_addr
+            else [target_pair.fee]
+        )
+
+        fees = [int(fee) for fee in fees] # TODO fix, shouldn't need to do this
+
+        price_helper = self.get_price_helper(target_pair)
+        token_raw_received = price_helper.get_amount_out(
+            amount_in=reserve_raw,
+            path=path,
+            fees=fees
+        )
+
+        token_received = target_pair.base.convert_to_decimal(token_raw_received)
+        
+        fee = self.get_pair_fee_multiplier(ts, pair)
+        assert fee is not None, "Uniswap v3 fee data missing"
+
+        price = float(reserve / token_received)
+
+        lp_fee = float(reserve) * fee
+
+        # TODO: Verify calculation
+        if intermediate_pair is not None:        
+            mid_price = price * (1 - self.get_pair_fee_multiplier(ts, intermediate_pair) - self.get_pair_fee_multiplier(ts, target_pair))
+        else:
+            mid_price = price * (1 - fee)
+
+        assert price >= mid_price, f"Bad pricing: {price}, {mid_price}"
+
+        return TradePricing(
+            price=float(price),
+            mid_price=float(mid_price),
+            lp_fee=lp_fee,
+            pair_fee=fees,
```

Comment:
Make sure `TradingPrice` data class decription types and comments are reflected to update the fact that this could be an a list, under which scenarios it is a list and what is the list order.

Also add a test where it checks the list is correctly read